DailyKit/CallManager/Models/CallParticipants.swift

```diff
@@ -0,0 +1,133 @@
+
+        var activeSpeaker: CallParticipant? {
+            didSet {
+                if let activeSpeaker, activeSpeaker.isLocal == false {
+                    speakerOrder.removeAll(where: { $0 == activeSpeaker.id })
+                    speakerOrder.append(activeSpeaker.id)
+                    speakerOrder = Array(speakerOrder.prefix(limit))
+                }
+            }
+        }
+
+        func build() -> CallParticipants {
+            var ids: Set<CallParticipant.ID> = []
+            var newVisible: [Int: CallParticipant] = [:]
+
+            func append(_ participant: CallParticipant) {
+                guard ids.count < limit else { return }
+                guard ids.contains(participant.id) == false else { return }
+
+                ids.insert(participant.id)
+                newVisible[newVisible.count] = participant
+            }
+
+            func append(_ participantIDs: [CallParticipant.ID]) {
+                var withVideo: [CallParticipant] = []
+                var withoutVideo: [CallParticipant] = []
+
+                let participants = participantIDs
+                    .map { remote[$0] }
+                    .compactMap { $0 }
+
+                for participant in participants {
+                    if participant.hasVideo {
+                        withVideo.append(participant)
+                    } else {
+                        withoutVideo.append(participant)
+                    }
+                }
+
+                withVideo.forEach { append($0) }
+                withoutVideo.forEach { append($0) }
+            }
+
+            if let activeSpeaker, activeSpeaker.isLocal == false {
+                append(activeSpeaker)
+            }
+
+            append(speakerOrder.reversed())
+            append(joinOrder.reversed())
+
+            return CallParticipants(
+                local: local,
+                remote: remote,
+                visible: newVisible
+            )
+        }
+
+        mutating func handleJoined(_ participant: CallParticipant) {
+            assert(participant.isLocal == false, "Unexpected join event for the local participant.")
+
+            joinOrder.append(participant.id)
+            joinOrder = Array(joinOrder.prefix(limit))
+
+            remote[participant.id] = participant
+        }
+
+        mutating func handleUpdated(_ participant: CallParticipant) {
+            if participant.isLocal {
+                local = participant
+            } else {
+                remote[participant.id] = participant
+            }
+        }
+
+        mutating func handleLeft(_ participant: CallParticipant) {
+            assert(participant.isLocal == false, "The local participant cannot be removed.")
+
+            joinOrder.removeAll { $0 == participant.id }
+            speakerOrder.removeAll { $0 == participant.id }
+
+            remote.removeValue(forKey: participant.id)
+        }
+    }
+
+    public var count: Int { 1 + remote.count }
```

Comment:
Nit: wondering if for readability it would be nicer to move these property definitions to the top of the class (but since I notice you sometimes put these in other locations in the class I'm starting to suspect Swift convention is different)

Reply:
+1!

Same as https://github.com/daily-demos/daily-ios-starter-kit/pull/1#discussion_r1255610183

Reply:
See my comment on #1 too. I wanted to have the `Builder` type above the properties for `CallParticipants` itself. It's gotten a lot bigger though, so I'll move the `Builder` to its own file.